crates/loader/src/validation.rs

```diff
@@ -1,23 +1,276 @@
-        {
-            return Ok(());
+    parse_allowed_http_hosts(http_hosts).map(|_| ())
+}
+
+// Parses a list of allowed HTTP hosts
+pub fn parse_allowed_http_hosts(raw: &Option<Vec<String>>) -> Result<AllowedHttpHosts> {
+    match raw {
+        None => Ok(AllowedHttpHosts::AllowSpecific(vec![])),
+        Some(list) => {
+            if list
+                .iter()
+                .any(|domain| domain == wasi_outbound_http::ALLOW_ALL_HOSTS)
+            {
+                Ok(AllowedHttpHosts::AllowAll)
+            } else {
+                let parse_results = list
+                    .iter()
+                    .map(|h| parse_allowed_http_host(h))
+                    .collect::<Vec<_>>();
+                let (hosts, errors) = partition_results(parse_results);
+
+                if errors.is_empty() {
+                    Ok(AllowedHttpHosts::AllowSpecific(hosts))
+                } else {
+                    Err(anyhow!(
+                        "One or more allowed_http_hosts entries was invalid:\n{}",
+                        errors.join("\n")
+                    ))
+                }
+            }
+        }
+    }
+}
+
+fn parse_allowed_http_host(text: &str) -> Result<AllowedHttpHost, String> {
+    // If you call Url::parse, it accepts things like `localhost:3001`, inferring
+    // `localhost` as a scheme. That's unhelpful for us, so we do a crude check
+    // before trying to treat the string as a URL.
+    if text.contains("//") {
+        parse_allowed_http_host_as_schemed_url(text)
+    } else {
+        parse_allowed_http_host_as_host(text)
+    }
+}
+
+fn parse_allowed_http_host_as_host(text: &str) -> Result<AllowedHttpHost, String> {
+    match text.split(':').collect::<Vec<_>>()[..] {
+        [host] => match is_valid_host_name(host) {
+            true => Ok(AllowedHttpHost::host(host)),
+            false => Err(format!("{} isn't a valid host or host:port string", text)),
+        },
+        [host, port] => match (is_valid_host_name(host), u16::from_str(port)) {
+            (true, Ok(p)) => {
+                if is_allowed_port(p) {
+                    Ok(AllowedHttpHost::host_and_port(host, p))
+                } else {
+                    Err(format!(
+                        "Can't allow {} because the port is reserved for non-HTTP protocols",
+                        text
+                    ))
+                }
+            }
+            _ => Err(format!("{} isn't a valid host or host:port string", text)),
+        },
+        _ => Err(format!("{} isn't a valid host or host:port string", text)),
+    }
+}
+
+fn is_valid_host_name(text: &str) -> bool {
+    return !text.is_empty()
+        && text.len() <= 253
+        && text
+            .chars()
+            .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '-' || c == '.');
+}
+
+fn parse_allowed_http_host_as_schemed_url(text: &str) -> Result<AllowedHttpHost, String> {
```

Comment:
I think this would be easier to follow in straight procedural style, e.g.

```rust
let url = Url::parse(text)
    .map_err(|e| format!("{text} isn't a valid HTTP host URL: {e}"))?;

let host = url.host_str();
if host.is_none() {
    return Err(format!("{text} doesn't contain a host name"));
}
...
```

Perhaps just personal preference. :shrug: